salt/cloud/clouds/ec2.py

```diff
@@ -1223,13 +1223,16 @@ def _get_subnetname_id(subnetname):
         opts=__opts__,
         sigver="4",
     ):
-        tags = subnet.get("tagSet", {}).get("item", {})
-        if not isinstance(tags, list):
-            tags = [tags]
-        for tag in tags:
-            if tag["key"] == "Name" and tag["value"] == subnetname:
-                log.debug("AWS Subnet ID of %s is %s", subnetname, subnet["subnetId"])
-                return subnet["subnetId"]
+        if "tagSet" in subnet:
+            tags = subnet.get("tagSet", {}).get("item", {})
+            if not isinstance(tags, list):
+                tags = [tags]
+            for tag in tags:
+                if tag["key"] == "Name" and tag["value"] == subnetname:
```

Comment:
This is the line that's raising the exception. I'm not sure what is *in* tag... but clearly it was missing a `"key"`.

A fix that *seems* correct (without looking into it deeper, I'm not 100% confident) would be to change this:

```
if tag.get('key') == "Name" and tag.get("value") == subnetname:
```

This does assume that the subnetname is going to be non-None, which may be incorrect. If `subnetname` could ever be none, `tag.get("value", object()) == subnetname` would be better, since subnetname will always be something not a brand new object.

Reply:
@waynew I've debugged this using `--log-level=trace` and if there's no tags whatsoever, you end up with the complete absence of `tagSet` completely.   The AWS data model seems to just not return tagSet at all if you have no tags whatsoever.  So the solution proposed would end up trying to do a `.get` on an empty dictionary `{}` yielding a similar error.

Alternatively we can look for the `tagSet` key, but because of `pep8` standard it doesn't like the `.has_key()` so ended up having to use `in`

Hope that makes sense, I've tested this fix and it works in skipping the subnets that are missing the `tagSet` key completely (which is an okay assumption, if you have no `tagSet`, you then don't have a `Name` tag).  Generally the AWS defaults don't come with `Name` tags.

Reply:
Oh, haha. This is exactly why we require writing tests. I just threw together a quick example to try and illustrate how they're the same...

```
def as_written(subnet, subnetname='bob'):
    tags = subnet.get("tagSet", {}).get("item", {})
    if not isinstance(tags, list):
        tags = [tags]
    for tag in tags:
        assert False, "We had tags, what?"
        if tag["key"] == "Name" and tag["value"] == subnetname:
            return subnet["subnetId"]


def as_changed(subnet, subnetname='bob'):
    if "tagSet" in subnet:
        tags = subnet.get("tagSet", {}).get("item", {})
        if not isinstance(tags, list):
            tags = [tags]
        for tag in tags:
            assert False, "We had tags, what?"
            if tag["key"] == "Name" and tag["value"] == subnetname:
                return subnet["subnetId"]


data = {"no tagSet here": "sersiously, there isn't one"}
as_written(data)
as_changed(data)

other_data = {"tagSet": "there is a tagset, but no key in any of these things"}

try:
    as_written(other_data)
except AssertionError:
    print('oops as written does not work')

try:
    as_changed(other_data)
except AssertionError:
    print('oops as changed does not work')
```

Except... they're not. Because a little check:

```
def as_written(subnet, subnetname='bob'):
    tags = subnet.get("tagSet", {}).get("item", {})
    if not isinstance(tags, list):
        print('lol, empty dict is not a list')
        tags = [tags]
        print('here are tags:', tags)
```

What's happening is that when tagSet does not exist, neither does item, so the *default* value is `{}`. And `{}` isn't an instance of list, doh!

It looks like the correct fix is just 

```
    tags = subnet.get("tagSet", {}).get("item", {})
    # to this ----------------------------------^v
    tags = subnet.get("tagSet", {}).get("item", [])
```

Subtle change, but that's the kind of thing that gets exposed when tests are written to pass this function:

1. an empty dictionary
2. a dictionary with "tagSet", but is empty
3. a dictionary with tagSet but the item is not a list (I guess it would be just a single dict with key and value
4. a dictionary with tagSet and the item is a list of items with key/value

And of course, the subnet is gonna need "subnetId" when the key/value matches.

If you need more help - attend one of the [Test Clinics on Twitch](https://www.twitch.tv/saltstackinc/schedule) and I'll be glad to clarify any questions! (Also, they are recorded so if you're unable to make that time, let me know and I can also try to record something a little more async.

HTH!

Reply:
@waynew thanks I went ahead and tested, both solutions work (what I had written and yours).  If this is going to block the fix I'll just swap to yours.

Maybe we should just get rid of that check altogether, if we go with what I added to reduce confusion?  I just didn't want to touch it unnecessarily.  So it looks like it's creating an empty tagSet `{}` with a tag empty tag array `[]` that will just get bypassed in the next loop.

```python
        tags = subnet.get('tagSet', {}).get('item', {})
        if not isinstance(tags, list):
            tags = [tags]
```

But what I have works as well, it just completely bypasses the case where there is no `tagSet`, here's an example output from the AWS `DescribeSubnets` (with sensitive data masked).  The first one has no name, so `tagSet` completely doesn't exist so doing `if "tagSet" in subnet:` will just bypass that case.

```xml
        <item>
            <subnetId>subnet-*******</subnetId>
            <state>available</state>
            <vpcId>vpc-********</vpcId>
            <cidrBlock>********</cidrBlock>
            <availableIpAddressCount>250</availableIpAddressCount>
            <availabilityZone>********</availabilityZone>
            <defaultForAz>false</defaultForAz>
            <mapPublicIpOnLaunch>false</mapPublicIpOnLaunch>
        </item>
        <item>
            <subnetId>subnet-********</subnetId>
            <state>available</state>
            <vpcId>vpc-********</vpcId>
            <cidrBlock>********</cidrBlock>
            <availableIpAddressCount>245</availableIpAddressCount>
            <availabilityZone>********</availabilityZone>
            <defaultForAz>false</defaultForAz>
            <mapPublicIpOnLaunch>false</mapPublicIpOnLaunch>
            <tagSet>
                <item>
                    <key>Name</key>
                    <value>Subnet with a Name</value>
                </item>
            </tagSet>
        </item>
```

Salt-cloud has been great, just really want get this fixed so it no longer blocks us from using `subnetname` rather than having to hardcode and create mappings to `SubnetId` in our IAC.   Will make our automation a lot smoother!

Reply:
Actually went ahead and incorporated your suggestion as well @waynew since there's probably other corner cases that this covers & that current line is broken anyways.